File: src/main/java/me/sul/worldmanager/summonmob/MobCleaner.java
package me.sul.worldmanager.summonmob;

import net.minecraft.server.v1_12_R1.EntityMonster;
import org.bukkit.World;
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftEntity;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Monster;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.server.PluginDisableEvent;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.ArrayList;
import java.util.List;

public class MobCleaner implements Listener {
    private final Plugin plugin;

    private final List<World> activeWorlds = new ArrayList<>();
    private final int CLEAN_MOB_PERIOD;


    public MobCleaner(Plugin plugin, String parentNode, List<World> activeWorlds) {
        this.plugin = plugin;
        this.activeWorlds.addAll(activeWorlds);
        this.CLEAN_MOB_PERIOD = plugin.getConfig().getInt(parentNode + ".cleaning-mob-period") * 20;

        registerCleanUselessMobScheduler();
    }
    private void registerCleanUselessMobScheduler() {
        new BukkitRunnable() {
            @Override
            public void run() {
                cleanAutoSummonableMob(false);
            }
        }.runTaskTimer(plugin, 0, CLEAN_MOB_PERIOD);
    }

    @EventHandler
    public void onDisable(PluginDisableEvent e) {
        cleanAutoSummonableMob(true);
    }

    private void cleanAutoSummonableMob(boolean includeUsefulMobs) {
        for (World world : activeWorlds) {
            // mobType은 Skeleton, Zombie여야 함
            for (Entity entity : world.getEntities()) {
                if (entity instanceof Monster && entity.hasMetadata(MobSummoner.AUTO_SUMMONED_MOB_METAKEY)) {
                    if (includeUsefulMobs || ((EntityMonster) ((CraftEntity)entity).getHandle()).getGoalTarget() == null) {
                        entity.remove();
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/kr/sul/worldmanager/summonmob/MobCleaner.kt
package kr.sul.worldmanager.summonmob

import net.minecraft.server.v1_12_R1.EntityMonster
import org.bukkit.World
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftEntity
import org.bukkit.entity.Monster
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.server.PluginDisableEvent
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitRunnable

class MobCleaner(private val plugin: Plugin, parentNode: String, activeWorldsParam: List<World>) : Listener {
    companion object {
        val activeWorlds = ArrayList<World>()
        var CLEANING_MOB_PERIOD: Int = 999999999
    }
    init {
        activeWorlds.addAll(activeWorldsParam)
        CLEANING_MOB_PERIOD = plugin.config.getInt("$parentNode.cleaning-mob-period") * 20
        registerCleanUselessMobScheduler()
    }

    private fun registerCleanUselessMobScheduler() {
        object : BukkitRunnable() {
            override fun run() {
                cleanAutoSummonableMob(false)
            }
        }.runTaskTimer(plugin, 0, CLEANING_MOB_PERIOD.toLong())
    }

    @EventHandler
    fun onDisable(e: PluginDisableEvent) {
        cleanAutoSummonableMob(true)
    }

    private fun cleanAutoSummonableMob(toAllMobs: Boolean) {
        for (world in activeWorlds) {
            world.entities.stream()
                    .filter { it is Monster && it.hasMetadata(MobSummoner.AUTO_SUMMONED_MOB_METAKEY) }
                    .filter { toAllMobs || ((it as CraftEntity).handle as EntityMonster).goalTarget == null }
                    .forEach { it.remove() }
        }
    }
}